rfcs/0164-bootspecv2.md

```diff
@@ -0,0 +1,300 @@
+  flexibility for other ecosystems.
+- Enhance initrd flexibility, allowing developers to optimize their systems by
+  supporting multiple initrds, including one for microcode and other specific
+  purposes.
+
+### Non-Goals
+[non-goals]: #non-goals
+
+- Store TPM2-related information (hashes)
+  We believe that Bootspec is still too immature for this and pcrlock
+  (https://github.com/systemd/systemd/pull/28891) offers a more reliable and
+  robust solution for generating signed PCR policies.
+- Supporting SecureBoot.
+  Secure Boot has one maintained implementation that is being upstreamed: https://github.com/nix-community/lanzaboote
+  which was enabled by Bootspec v1.
+- Specifying how to discover generations. This is desirable, but should not be tied to bootspec directly since bootspec may be useful with diverse discovery mechanisms.
+
+# Proposed Solution
+[proposed-solution]: #proposed-solution
+
+- `initrd` will be removed from the v2
+- `initrds` will be introduced as a list of initrd (compressed or uncompressed
+  CPIO archives), this list can be empty, but the field is **required**
+  nonetheless.
+- `initrdSecrets` will now be a hashmap of strings (key) and strings (value)
+  where the key is the "name" of a secret and the value is the "path" towards a
+  secret.
+- `fdtdir` will be introduced as an opaque string to a directory in a shape as
+  expected by U-Boot extlinux `FDTDIR` directive — it is **optional**
+- `devicetree` will be introduced as a path to a single devicetree that will be
+  hardcoded — it is **optional**
+
+All the Bootspec ecosystem will be updated as part of this specification.
+
+### Bootspec Format v2
+[format-v2]: #format-v2
+
+
+Using the following JSON:
+
+```json5
+{
+  // Toplevel key describing the version of the specification used in the document
+  "org.nixos.bootspec.v2": {
+    // (Required) System type the bootspec is intended for (e.g. `x86_64-linux`, `aarch64-linux`)
+    "system": "x86_64-linux",
+
+    // (Required) Path to the stage-2 init, executed by the initrd (if present)
+    "init": "/nix/store/xxx-nixos-system-xxx/init",
+
+    // (Required) List of paths to the initrd, can be empty
+    "initrds": [ "/nix/store/xxx-initrd-linux/initrd" ],
+
+    // (Required) Hash map of desired secrets for that generation inside of the initrd.
+    // Implementors of a bootloader installation procedure should examine their options
+    // to securely make available the secret inside the initrd phase.
+    // This may involve leveraging TPM2 via systemd-credentials or any measure you deem
+    // to be reasonable in the context.
+    // The legacy behavior is to prepare a CPIO archive for each file and
+    // extend the `initrds` fields with those CPIO archives.
+    // Make sure the location where the secrets are dropped in the initrd are visible
+    // for the user.
+    "initrdSecrets": {
+      "my-private-key": "/etc/nixos/secrets/wireguard-private-key",
+    }
```

Comment:
I wonder if the various implementations possible with initrd secrets makes this design too specific for the design space. What if this mechanism was split out into an "extension" like specifications? I could imagine an initrdSecrets variant here which is pulling secrets from Vault, or AWS Secrets Manager, or `pass`.

Reply:
I touch on the matter in the FAQ, I believe that it's up to the bootloader implementation to offer another backend for interpreting the values of the `initrdSecrets` if needed. Some trivial secret engine could even offer a filesystem to mount the secrets and let them be downloaded/offered at runtime.